ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,297 @@
+                parent = await resource.get_parent()
+                children = await parent.get_children(
+                    r_filter=ResourceFilter(
+                        resource.get_most_specific_tags(),
+                        attribute_filters=ResourceAttributeValueFilter(
+                            attribute=attribute,
+                            value=attribute_value)
+                        )
+                    )
+                if len(children) > 1:
+                    attribute_collisions[attribute.__name__] = attribute_value
+                    continue
+                return attribute, attribute_value
+        if len(attribute_collisions) == 0:
+            raise SelectableAttributesError(
+                f"Resource with ID {resource.get_id()} does not have a selectable attribute."
+            )
+        else:
+            msg = []
+            for collision, value in attribute_collisions.items(): 
+                msg.append(f"Resource with ID {resource.get_id()} cannot be uniquely identified by attribute {attribute.__name__} (resource has value {attribute_value}).")
+            raise SelectableAttributesError("\n".join(msg))
+
+    async def _generate_name(self, resource: Resource) -> str:
+        root_resource = await self._get_root_resource(resource)
+        most_specific_tag = list(resource.get_most_specific_tags())[0].__name__.lower()
+        _, selectable_attribute_value = await self._get_selectable_attribute(resource)
+        name = f"{most_specific_tag}_{selectable_attribute_value}"
+        name = re.sub(r"[\s\-\.\/\\]", "_", name)
+        if name in self.script_sessions[root_resource.get_id()].resource_variable_names.values():
+            parent = await resource.get_parent()
+            return f"{self.script_sessions[root_resource.get_id()].resource_variable_names[parent.get_id()]}_{name}"
+        return name
+
+    def _get_session(self, resource_id: bytes) -> ScriptSession:
+        session = self.script_sessions.get(resource_id, None)
+        if session is None:
+            session = ScriptSession()
+            self.script_sessions[resource_id] = session
+
+        return session
+
+    def _get_script(self, resource_id: bytes, target_type: Optional[ActionType] = None) -> str:
+        script = []
+        script.append(self.script_sessions[resource_id].boilerplate_header)
+        for script_action in self.script_sessions[resource_id].actions:
+            # Always include UNDEF actions like variable assignments
+            if (
+                target_type is None
+                or script_action.action_type == target_type
+                or script_action.action_type == ActionType.UNDEF
+            ):
+                script.append(f"{script_action.action}")
+        script.append(self.script_sessions[resource_id].boilerplate_footer)
+        script = "\n".join(script)
+        script = self._dedent(script)
+        res = format_str("\n".join(script), mode=FileMode())
```

Comment:
The script is already joined?

```suggestion
        res = format_str(script, mode=FileMode())
```

Reply:
`dedent` expects a string and returns a list. The second join could be moved inside of `dedent` so that it takes a string and returns a string, although that's just shifting the operation around rather than removing it. I did a quick experiment and it doesn't appear that it's straightforward to remove the first join and have `dedent` take a list. The reason being is that the the `splitlines` call in `dedent` removes '\n' chars that are present in the `script` list before the first call to `join`, so the two lists are not identical.

Reply:
`dedent` expects a string and returns a list. The second join could be moved inside of `dedent` so that it takes a string and returns a string, although that's just shifting the operation around rather than removing it. I did a quick experiment and it doesn't appear that it's straightforward to remove the first join and have `dedent` take a list. The reason being is that the the `splitlines` call in `dedent` removes '\n' chars that are present in the `script` list before the first call to `join`, so the two lists are not identical.

Reply:
No need for `dedent` to take a list. I do think it would be cleaner if `dedent` took a string and returned a string. That interface just feels a little more natural to me.